components/ee/ws-scheduler/pkg/scaler/driver.go

```diff
@@ -0,0 +1,380 @@
+		Client:     api.NewWorkspaceManagerClient(conn),
+		conn:       conn,
+		Controller: controller,
+		stop:       make(chan struct{}),
+	}, nil
+}
+
+// WorkspaceManagerPrescaleDriver implements a prescale driver using ws-manager's ghost pods
+type WorkspaceManagerPrescaleDriver struct {
+	Config WorkspaceManagerPrescaleDriverConfig
+
+	Client api.WorkspaceManagerClient
+	conn   *grpc.ClientConn
+
+	Controller Controller
+
+	stop chan struct{}
+	once sync.Once
+}
+
+type workspaceStatus struct {
+	Count              WorkspaceCount
+	DeletionCandidates []string
+}
+
+// Run runs the prescale driver until Stop() is called
+func (wspd *WorkspaceManagerPrescaleDriver) Run() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	statusChan := make(chan workspaceStatus)
+	go func() {
+		for {
+			err := wspd.maintainWorkspaceStatus(ctx, statusChan)
+			if err == context.Canceled {
+				return
+			}
+			if err != nil {
+				log.WithError(err).Error("cannot maintain workspace count")
+				time.Sleep(1 * time.Second)
+			}
+		}
+	}()
+
+	var renewal <-chan time.Time
+	if wspd.Config.Renewal.Percentage > 0 && wspd.Config.Renewal.Interval > 0 {
+		t := time.NewTicker(time.Duration(wspd.Config.Renewal.Interval))
+		defer t.Stop()
+		renewal = t.C
+	}
+	houseKeeping := time.NewTicker(1 * time.Minute)
+	defer houseKeeping.Stop()
+
+	var (
+		counts         = make(chan WorkspaceCount)
+		status         workspaceStatus
+		startingGhosts = make(map[string]time.Time)
+	)
+
+	delta := wspd.Controller.Control(ctx, counts)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case status = <-statusChan:
+			for _, id := range status.DeletionCandidates {
+				delete(startingGhosts, id)
+			}
+			status.Count.Ghost += len(startingGhosts)
+			counts <- status.Count
+			log.WithField("counts", counts).Debug("status update")
+		case <-houseKeeping.C:
+			for id, t1 := range startingGhosts {
+				if time.Since(t1) < maxGhostStartTime {
+					continue
+				}
+				delete(startingGhosts, id)
+				status.Count.Ghost--
+			}
+		case <-renewal:
+			d := int(float64(len(status.DeletionCandidates)) * float64(wspd.Config.Renewal.Percentage) * 0.01)
+			if d == 0 {
+				continue
+			}
+			if d > len(status.DeletionCandidates) {
+				d = len(status.DeletionCandidates)
+			}
+			log.WithField("count", d).Info("attempting to renew ghost workspaces")
+			err := wspd.stopGhostWorkspaces(ctx, status.DeletionCandidates[:d])
+			if err != nil {
+				log.WithError(err).Error("cannot stop ghost workspaces during renewal")
+				continue
+			}
+			ids, err := wspd.startGhostWorkspaces(ctx, d, status)
+			if err != nil {
+				log.WithError(err).Error("cannot start ghost workspaces during renewal")
+				continue
+			}
+			for _, id := range ids {
+				startingGhosts[id] = time.Now()
+				status.Count.Ghost++
+			}
+		case d := <-delta:
+			if d == 0 {
+				continue
+			}
+			if wspd.Config.ReactionDelay > 0 {
+				time.Sleep(time.Duration(wspd.Config.ReactionDelay))
```

Comment:
Could you elaborate on the reason for this? Would it make sense to refresh `status` after such a delay and before acting on it (even defensively?).

Reply:
I've added a comment here.

We could, instead of just delaying everything, implement a scheme we where we collect updates and if there hasn't been an update in the last X time, or we've reached N updates (after which we'll forcefully wait X time), run the controller. This would probably make the controller less "hectic".

Reply:
I'd even argue for a approach where we'd constantly update the status but only call the controller in a fixed interval. That would make it far easier to reason about it, plus it's easier to guarantee that it is stable (not "hectic") as opposed to escalating itself.